R/response_process.R

```diff
@@ -0,0 +1,165 @@
+#' Process a Google API response
+#'
+#' @description
+#' `response_process()` is intended primarily for internal use in client
+#' packages that provide high-level wrappers for users. Typically applied as the
+#' final step in this sequence of calls:
+#'   * Request prepared with [request_build()].
+#'   * Request made with [request_make()].
+#'   * Response processed with [response_process()].
+#'
+#' All that's needed for a successful request is to parse the JSON extracted via
+#' `httr::content()`. Therefore, the main point of `response_process()` is to
+#' handle less happy outcomes:
+#'   * Status codes in the 400s (client error) and 500s (server error). The
+#'     structure of the error payload varies across Google APIs and we try to
+#'     create a useful message for all variants we know about.
+#'   * Non-JSON content type, such as HTML.
+#'   * Status code in the 100s (information) or 300s (redirection). These are
+#'     unexpected.
+#'
+#' @details
+#' If `process_response()` results in an error, a redacted version of the `resp`
+#' input is returned in the condition (auth tokens are removed). Use functions
+#' such as `rlang::last_error()` or `rlang::catch_cnd()` to capture the
+#' condition and do a more detailed forensic examination.
+#'
+#' The `response_as_json()` helper is exported only as an aid to maintainers who
+#' wish to use their own `error_message` function, instead of gargle's built-in
+#' `gargle_error_message()`. When implementing a custom `error_message`
+#' function, call `response_as_json()` immediately on the input in order to
+#' inherit gargle's handling of non-JSON input.
+#'
+#' @param resp Object of class `response` from [httr].
+#' @param error_message Function that produces an informative error message from
+#'   the primary input, `resp`. It must return a character vector.
+#'
+#' @return The content of the request, as a list. An HTTP status code of 204 (No
+#'   content) is a special case returning `TRUE`.
+#' @family requests and responses
+#' @export
+response_process <- function(resp, error_message = gargle_error_message) {
+  code <- httr::status_code(resp)
+
+  if (code >= 200 && code < 300) {
+    if (code == 204) {
+      # HTTP status: No content
+      TRUE
+    } else {
+      response_as_json(resp)
+    }
+  } else {
+    stop_request_failed(error_message(resp), resp)
```

Comment:
The one thing we lost in the refactor is that, when we error on a status in the 100s or 300s, we don't highlight that particular cause. And I imagine the message will be very minimal, as the content probably isn't really one of the errors I've planned around. But I also think this situation is OK until someone tells us it's not.